modules/light-clients/08-wasm/light_client_module.go

```diff
@@ -257,6 +258,53 @@ func (lcm LightClientModule) RecoverClient(ctx sdk.Context, clientID, substitute
 // the new client state, consensus state, and any other client metadata.
 //
 // CONTRACT: clientID is validated in 02-client router, thus clientID is assumed here to have the format 08-wasm-{n}.
-func (LightClientModule) VerifyUpgradeAndUpdateState(ctx sdk.Context, clientID string, newClient []byte, newConsState []byte, upgradeClientProof, upgradeConsensusStateProof []byte) error {
-	return nil
+func (lcm LightClientModule) VerifyUpgradeAndUpdateState(
+	ctx sdk.Context,
+	clientID string,
+	newClient []byte,
+	newConsState []byte,
+	upgradeClientProof,
+	upgradeConsensusStateProof []byte,
+) error {
+	var (
+		cdc               = lcm.keeper.Codec()
+		newClientState    exported.ClientState
+		newConsensusState exported.ConsensusState
+	)
+
+	if err := cdc.UnmarshalInterface(newClient, &newClientState); err != nil {
+		return err
+	}
+	newWasmClientState, ok := newClientState.(*types.ClientState)
+	if !ok {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClient, "expected client state type %T, got %T", (*types.ClientState)(nil), newClientState)
+	}
+	if err := newWasmClientState.Validate(); err != nil {
+		return err
+	}
+
+	if err := cdc.UnmarshalInterface(newConsState, &newConsensusState); err != nil {
+		return err
+	}
+	newWasmConsensusState, ok := newConsensusState.(*types.ConsensusState)
+	if !ok {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidConsensus, "expected consensus state type %T, got %T", (*types.ConsensusState)(nil), newConsensusState)
+	}
+	if err := newWasmConsensusState.ValidateBasic(); err != nil {
+		return err
+	}
```

Comment:
This code needs to be aligned with 07-tendermint impl, shame that this is not caught with tests.. hmmm